File: src/js-api/geo-core/src/core/layer.js
// TODO add SRID .. on call 

Ext.ns("Geo.core");
Geo.core.Layer = Ext.extend(Ext.util.Observable, {
	constructor: function (config) {
		this.addEvents(
			'query',
			'metadata',
			'bbox',
			'staticbbox'
		);
		Geo.core.Layer.superclass.constructor.call(config);
	}
	,
	query:function(layer, fields, criteria, paging, start, limit, wkt){
		paging = paging || false;
		start = start || 0;
		limit = limit || 0;
		wkt = wkt || false; 
		
		if(!criteria)
			criteria = [];
		
		var p = {
			paging : paging,
			start : start,
			limit : limit,
			layer : layer,
			fields : fields,
			criteria : criteria,
			wkt:wkt
		};

		Ext.Ajax.request({
			url: Geo.UriTemplate.getUri('layerQuery'),
			method: 'Post',
			headers: { 'Content-Type': 'text/json' },
			jsonData: p,
			scope: this,
			success: function (response, options) {
				var result = Ext.util.JSON.decode(response.responseText);
				this.fireEvent('query', result);
			}
		});
	}
	,
	getMetadata:function(layerName){
		
		var p = {
			layerName : layerName
		};
		
		Ext.Ajax.request({
			url: Geo.UriTemplate.getUri('layerMetadata','?' +  Ext.urlEncode(p)),
			method: 'Get',
			headers: { 'Content-Type': 'text/json' },
			scope:this,
			success: function (response, options) {
				var result = Ext.util.JSON.decode(response.responseText);
				this.fireEvent('metadata', result);
			}
		});
	}
	,
	getBBox:function(layerName, srid){
		
		var srid = srid || '96';
		
		var p = {
			layerName : layerName,
			srid : srid
		};
		
		Ext.Ajax.request({
			url: Geo.UriTemplate.getUri('layerBbox','?' +  Ext.urlEncode(p)),
			method: 'Get',
			headers: { 'Content-Type': 'text/json' },
			scope:this,
			success: function (response, options) {
				var result = Ext.util.JSON.decode(response.responseText);
				this.fireEvent('bbox', result);
			}
		});
	}
	,
	getStaticBBox:function(layerName){
		
		var p = {
			layerName : layerName
		};
		
		Ext.Ajax.request({
			url: Geo.UriTemplate.getUri('layerStaticBbox', '?' +  Ext.urlEncode(p)),
			method: 'Get',
			headers: { 'Content-Type': 'text/json' },
			scope:this,
			success: function (response, options) {
				var result = Ext.util.JSON.decode(response.responseText);
				this.fireEvent('staticbbox', result);
			}
		});
	}
});

File: src/server/lib/geo/rest/layer.py
from bottle import *
from geo.model import *
import geo.rest.vo as vo
import json

@get('/layer/metadata')
def metadata():
	layer_name = request.GET.get('layerName')	
	l = Layer(layer_name)
	metadatas = map(lambda m: vo.metadata(m), l.metadata())
	return vo.collection(metadatas, len(metadatas))
	
@post('layer/query')
def query():
	o = json.load(request.body)
	layer_name = o['layer']
	fields = o['fields']
	criteria = o['criteria']
	paging = o['paging']
	start = o['start']
	limit = o['limit']
	wkt = o['wkt']
	
	l = Layer(layer_name)
	results = l.query(fields, criteria, paging, start, limit,True, wkt)
	return vo.collection(results, l.query_count(criteria))


@get('/layer/bbox')
def bbox():
	layer_name = request.GET.get('layerName')
	srid = request.GET.get('srid')
	l = Layer(layer_name,32719)
	bbox = l.bbox(96)
	return vo.bbox(bbox)

@get('/layer/staticbbox')
def staticbbox():
	layer_name = request.GET.get('layerName')
	l = Layer(layer_name)
	bbox = l.static_bbox()
	return vo.bbox(bbox)